File: crates/ward_runtime/py/wardscript/mock.py
"""A deterministic model for tests: scripted answers per `ai fn`."""

from __future__ import annotations

import json
from typing import Any, Mapping

from .errors import WardError
from .model import AiRequest
from .schema import encode


class MockError(WardError):
    pass


class Raw:
    """An answer given as JSON text as-is, e.g. to test invalid output."""

    def __init__(self, text: str) -> None:
        self.text = text


class Seq:
    """Answers given one per call, in order. Running out is an error."""

    def __init__(self, *answers: Any) -> None:
        self.answers = list(answers)


class MockModel:
    """Answers each `ai fn` from `answers`, keyed by function name. An answer is a
    value (encoded as JSON, so records and enums work), a `Raw` text, a `Seq` of
    answers, or a callable taking the `AiRequest` and returning one of those."""

    def __init__(self, answers: Mapping[str, Any] | None = None) -> None:
        self.answers = dict(answers or {})
        self.calls: list[AiRequest] = []
        self._next: dict[str, int] = {}

    @classmethod
    def from_json(cls, text: str) -> MockModel:
        """`{"fn_name": answer, ...}`, answers as JSON values."""
        answers = json.loads(text)
        if not isinstance(answers, dict):
            raise MockError("mock answers must be a JSON object keyed by function name")
        return cls(answers)

    def complete(self, request: AiRequest) -> str:
        self.calls.append(request)
        if request.function not in self.answers:
            raise MockError(f"the mock model has no answer for `{request.function}`")
        return self._text(request, self.answers[request.function])

    def _text(self, request: AiRequest, answer: Any) -> str:
        if isinstance(answer, Seq):
            n = self._next.get(request.function, 0)
            if n >= len(answer.answers):
                raise MockError(
                    f"the mock model ran out of answers for `{request.function}` "
                    f"after {len(answer.answers)}"
                )
            self._next[request.function] = n + 1
            return self._text(request, answer.answers[n])
        if isinstance(answer, Raw):
            return answer.text
        if callable(answer) and not isinstance(answer, type):
            return self._text(request, answer(request))
        return json.dumps(encode(answer))


__all__ = ["MockError", "MockModel", "Raw", "Seq"]

File: crates/ward_runtime/ts/src/mock.ts
// A deterministic model for tests: scripted answers per `ai fn`.

import { WardError } from "./errors.ts";
import type { AiRequest, Completion } from "./model.ts";
import { encode } from "./schema.ts";

export class MockError extends WardError {}

/** An answer given as JSON text as-is, e.g. to test invalid output. */
export class Raw {
  readonly text: string;
  constructor(text: string) {
    this.text = text;
  }
}

/** An answer with what it cost. */
export class Usage {
  readonly answer: unknown;
  readonly tokens: number | null;
  readonly cost: number | null;
  constructor(answer: unknown, tokens: number | null = null, cost: number | null = 0) {
    this.answer = answer;
    this.tokens = tokens;
    this.cost = cost;
  }
}

/** Answers in turn: the first call gets the first. */
export class Seq {
  readonly answers: unknown[];
  constructor(...answers: unknown[]) {
    this.answers = answers;
  }
}

/** Answers each `ai fn` from `answers`, keyed by function name: a value (encoded as
 * JSON), a `Raw` text, a `Seq`, a `Usage`, an `Error` to throw, or a function of the
 * request returning one of those. */
export class MockModel {
  readonly answers: Record<string, unknown>;
  readonly calls: AiRequest[] = [];
  private readonly next = new Map<string, number>();

  constructor(answers: Record<string, unknown> = {}) {
    this.answers = answers;
  }

  static fromJson(text: string): MockModel {
    return new MockModel(JSON.parse(text) as Record<string, unknown>);
  }

  complete(request: AiRequest): string | Completion {
    this.calls.push(request);
    if (!(request.function in this.answers)) {
      throw new MockError(`the mock model has no answer for \`${request.function}\``);
    }
    const answer = this.text(request, this.answers[request.function]);
    return typeof answer === "string" ? { text: answer, tokens: null, cost: 0 } : answer;
  }

  private text(request: AiRequest, answer: unknown): string | Completion {
    if (answer instanceof Error) throw answer;
    if (answer instanceof Seq) {
      const n = this.next.get(request.function) ?? 0;
      if (n >= answer.answers.length) {
        throw new MockError(`the mock model ran out of answers for \`${request.function}\` after ${answer.answers.length}`);
      }
      this.next.set(request.function, n + 1);
      return this.text(request, answer.answers[n]);
    }
    if (answer instanceof Usage) {
      const inner = this.text(request, answer.answer);
      return { text: typeof inner === "string" ? inner : inner.text, tokens: answer.tokens, cost: answer.cost };
    }
    if (answer instanceof Raw) return answer.text;
    if (typeof answer === "function") return this.text(request, (answer as (r: AiRequest) => unknown)(request));
    return JSON.stringify(encode(answer));
  }
}
